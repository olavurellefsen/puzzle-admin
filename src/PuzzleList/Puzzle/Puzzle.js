import React, { useState, useContext, useRef } from "react";
import MainContext from "../../Context";
import {
  PuzzleContainer,
  PuzzleForm,
  PuzzleFieldCaption,
  PuzzleField,
  PuzzleInputField,
  PuzzleRightArrow
} from "./Puzzle.style";
import gql from "graphql-tag";
import { Mutation } from "react-apollo";

export default (props) => {
  const [state, dispatch] = useContext(MainContext);
  let { id, intro_audiofile, character, summary } = props.puzzle;
  const [summaryField, setSummaryField] = useState(
    summary !== null ? summary : ""
  );
  const puzzleSummary = useRef(null);

  // If the parent component is updated (someone else has changed the value), then update the field here
  if (
    summary !== summaryField &&
    summary !== null &&
    puzzleSummary.current !== null
  ) {
    if (puzzleSummary.current.defaultValue === summaryField) {
      setSummaryField(summary);
    }
  }

  const UPDATE_PUZZLE = gql`
    mutation update_puzzle($id: Int!, $summary: String!) {
      update_puzzle(where: { id: { _eq: $id } }, _set: { summary: $summary }) {
        affected_rows
      }
    }
  `;

  const selected = id === props.currentPuzzleId;
  return (
    <PuzzleContainer
      onClick={() => {
        dispatch({
          type: "setCurrentPuzzle",
          puzzle_id: id,
          scene_id: state.currentScene
        });
      }}
      selected={selected}
      data-testid="PuzzleContainer" 
    >
      <Mutation mutation={UPDATE_PUZZLE}>
        {(updatePuzzleTitle, { data }) => (
          <PuzzleForm
            onSubmit={e => {
              e.preventDefault();
              updatePuzzleTitle({
                variables: { id: id, summary: summaryField }
              });
              puzzleSummary.current.blur();
            }}
          >
            <PuzzleFieldCaption>PUZZLE {id}</PuzzleFieldCaption>
            <PuzzleInputField
              ref={puzzleSummary}
              name="summary"
              type="text"
              value={summaryField}
              onChange={e => setSummaryField(e.target.value)}
            />
            <PuzzleFieldCaption>AUDIOFILE</PuzzleFieldCaption>
            <PuzzleField>
              {intro_audiofile !== null ? intro_audiofile : "-"}
            </PuzzleField>
            <PuzzleFieldCaption>CHARACTER</PuzzleFieldCaption>
            <PuzzleField>{character !== null ? character : "-"}</PuzzleField>
          </PuzzleForm>
        )}
      </Mutation>
      <PuzzleRightArrow selected={selected}>></PuzzleRightArrow>
    </PuzzleContainer>
  );
}
