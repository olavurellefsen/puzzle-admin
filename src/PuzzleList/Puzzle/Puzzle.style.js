import styled from "styled-components";

export const PuzzleContainer = styled.div`
  height: 200px;
  width: 240px;
  padding: 10px;
  margin: 20px;
  align-self: center;
  box-shadow: 0 4px 8px 0 rgba(0,0,0,${props => (props.selected ? 0.4 : 0.2)});
  transition: 0.3s;
  transition-timing-function: cubic-bezier(0, 1, 0.5, 1);
  &:hover {
    box-shadow: 0 8px 16px 0 rgba(0,0,0,0.6);
  }
  opacity: ${props => (props.selected ? 1 : 0.3)};
`;

export const PuzzleForm = styled.form`
  display: flex;
  flex-direction: column;
  padding: 2px 16px;
`;

export const PuzzleFieldCaption = styled.div`
  font-size: 12px;
  font-weight: normal;
  color: gray;
`;

export const PuzzleInputField = styled.input`
  font-size: 14px;
  padding: 0px 0px 10px 0px;
  background-color: rgba(0, 0, 0, 0);
  border: 0px;
`;

export const PuzzleTextAreaField = styled.textarea`
  font-size: 14px;
  padding: 0px 0px 10px 0px;
  background-color: rgba(0, 0, 0, 0);
  border: 0px;
`;

export const PuzzleRightArrow = styled.div`
  font-size: 60px;
  color: gray;
  text-align: right;
  transform: translate(50px,-100px);
  visibility: ${props => (props.selected ? 'visible' : 'hidden')}
`;