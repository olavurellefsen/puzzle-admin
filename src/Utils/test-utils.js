import React from "react";
import { render } from "@testing-library/react";
import { ApolloProvider } from "react-apollo";
import MainContextStore from "../MainContextStore";
import client from "../heyBreydLocalizationClient";

const AllTheProviders = ({ children }) => {
  return (
    <ApolloProvider client={client}>
      <MainContextStore>{children}</MainContextStore>
    </ApolloProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
