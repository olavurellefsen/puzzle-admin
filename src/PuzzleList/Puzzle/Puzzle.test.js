import React from "react";
import { render, cleanup } from 'test-utils';
import Puzzle from "./Puzzle";
afterEach(cleanup)

const puzzle = { id: 1, intro_audiofile: "dummyfile", character: "dummycharacer", summary: "Lorum ipsum" };
it("should render correctly", () => {
  
  const { container } = render(<Puzzle puzzle={puzzle}/>);
  expect(container).toMatchSnapshot();
});

it("should show the PuzzleContainer element", () => {
  const { queryByTestId } = render(<Puzzle puzzle={puzzle} />);
  expect(queryByTestId("PuzzleContainer")).toBeTruthy();
});