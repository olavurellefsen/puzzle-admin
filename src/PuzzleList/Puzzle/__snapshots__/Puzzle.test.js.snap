// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <div
    class="sc-bdVaJa dSJplu"
    data-testid="PuzzleContainer"
  >
    <form
      class="sc-bwzfXH iYKgxb"
    >
      <div
        class="sc-htpNat hccyqP"
      >
        G√ÅTUHEITI 
        1
      </div>
      <textarea
        class="sc-ifAKCX eZonoI"
        name="summary"
        rows="3"
      >
        Lorum ipsum
      </textarea>
      <div
        class="sc-htpNat hccyqP"
      >
        INTROTEKSTUR
      </div>
      <textarea
        class="sc-ifAKCX eZonoI"
        name="introtext"
        rows="3"
      >
        Intro text
      </textarea>
      <div
        class="sc-htpNat hccyqP"
      >
        LEIKARI
      </div>
      <input
        class="sc-bxivhb gsEoFf"
        name="character"
        type="text"
        value="dummycharacer"
      />
    </form>
    <div
      class="sc-EHOje chNjoN"
    >
      &gt;
    </div>
  </div>
</div>
`;
